tests/test_ast.py

```diff
@@ -0,0 +1,70 @@
+from m2cgen import ast
+
+
+def test_count_exprs():
+    assert ast.count_exprs(
+        ast.BinNumExpr(ast.NumVal(1), ast.NumVal(2), ast.BinNumOpType.ADD)
+    ) == 3
+
+    assert ast.count_exprs(
+        ast.ExpExpr(ast.NumVal(2))
+    ) == 2
+
+    assert ast.count_exprs(
+        ast.VectorVal([
+            ast.NumVal(2),
+            ast.TanhExpr(ast.NumVal(3))
+        ])
+    ) == 4
+
+    assert ast.count_exprs(
+        ast.IfExpr(
+            ast.CompExpr(ast.NumVal(2), ast.NumVal(0), ast.CompOpType.GT),
+            ast.NumVal(3),
+            ast.NumVal(4),
+        )
+    ) == 6
+
+    assert ast.count_exprs(ast.SubroutineExpr(ast.NumVal(1))) == 1
+
+
+def test_count_exprs_exclude_list():
+    assert ast.count_exprs(
+        ast.BinNumExpr(ast.NumVal(1), ast.NumVal(2), ast.BinNumOpType.ADD),
+        exclude_list={ast.BinExpr, ast.NumVal}
+    ) == 0
+
+    assert ast.count_exprs(
+        ast.BinNumExpr(ast.NumVal(1), ast.NumVal(2), ast.BinNumOpType.ADD),
+        exclude_list={ast.BinNumExpr}
+    ) == 2
+
+
+def test_count_all_exprs_types():
+    expr = ast.BinVectorNumExpr(
```

Comment:
Maybe even more complicated expr with deeper nesting? 🙂

Reply:
TBH, I don't see a value in having even deeper nesting since it won't really provide additional coverage. Additionally this will make the test more complex and somewhat obscure its purpose. Thanks for the feedback though.